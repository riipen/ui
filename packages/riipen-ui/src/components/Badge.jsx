import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

const RADIUS_STANDARD = 10;
const RADIUS_DOT = 5;

class Badge extends React.Component {
  static propTypes = {
    /**
     * The anchor of the badge.
     */
    anchor: PropTypes.shape({
      horizontal: PropTypes.oneOf(["left", "right"]).isRequired,
      vertical: PropTypes.oneOf(["bottom", "top"]).isRequired
    }),

    /**
     * The badge will be added relative to this node.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf(["negative", "positive", "primary", "secondary"]),

    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     */
    component: PropTypes.elementType,

    /**
     * The content rendered within the badge.
     */
    content: PropTypes.node,

    /**
     * Max count to show.
     */
    max: PropTypes.number,

    /**
     * Wrapped shape the badge should overlap.
     */
    overlap: PropTypes.oneOf(["circle", "rectangle"]),

    /**
     * Controls whether the badge is hidden when `badgeContent` is zero.
     */
    showZero: PropTypes.bool,

    /**
     * The variant to use.
     */
    variant: PropTypes.oneOf(["dot", "standard"])
  };

  static defaultProps = {
    anchor: {
      horizontal: "right",
      vertical: "top"
    },
    color: "primary",
    component: "span",
    max: 99,
    overlap: "rectangle",
    showZero: false,
    variant: "standard"
  };

  static contextType = ThemeContext;

  render() {
    const {
      anchor,
      children,
      classes,
      color,
      component: Component,
      content,
      max,
      overlap,
      showZero,
      variant
    } = this.props;

    if ((content === 0 && !showZero) || (content == null && variant !== "dot"))
      return null;

    const theme = this.context;

    const className = clsx("root");

    let value = "";

    if (variant !== "dot") {
      value = content > max ? `${max}+` : content;
    }

    return (
      <React.Fragment>
        <Component className={className}>
          {children}
          <span
            className={clsx(
              "badge",
              `${anchor.vertical}-${anchor.horizontal}-${overlap}`,
              color,
              variant,
              classes
            )}
          >
            {value}
          </span>
        </Component>
        <style jsx>{`
          .root {
            display: inline-flex;
            flex-shrink: 0;
            position: relative;
            vertical-align: middle;
          }

          .badge {
            align-content: center;
            align-items: center;
            border-radius: ${RADIUS_STANDARD}px;
            box-sizing: border-box;
            color: ${theme.palette.text.primary};
            display: flex;
            flex-direction: row;
            flex-wrap: wrap;
            font-family: ${theme.typography.fontFamily};
            font-size: 12px;
            font-weight: ${theme.typography.fontWeight.medium};
            height: ${RADIUS_STANDARD * 2}px;
            justify-content: center;
            line-height: 1;
            min-width: ${RADIUS_STANDARD * 2}px;
            padding: 0 6px;
            position: absolute;
          }

          .dot {
            height: ${RADIUS_DOT * 2}px;
            min-width: ${RADIUS_DOT * 2}px;
            padding: 0;
          }

          .primary {
            background-color: ${theme.palette.primary.main};
            color: ${theme.palette.primary.contrast};
          }
          .secondary {
            background-color: ${theme.palette.secondary.main};
            color: ${theme.palette.secondary.contrast};
          }
          .positive {
            background-color: ${theme.palette.positive.main};
            color: ${theme.palette.positive.contrast};
          }
          .negative {
            background-color: ${theme.palette.negative.main};
            color: ${theme.palette.negative.contrast};
          }

          .top-right-rectangle {
            top: 0;
            right: 0;
            transform: scale(1) translate(50%, -50%);
            transform-origin: 100% 0%;
          }
          .bottom-right-rectangle {
            bottom: 0;
            right: 0;
            transform: scale(1) translate(50%, 50%);
            transform-origin: 100% 100%;
          }
          .top-left-rectangle {
            top: 0;
            left: 0;
            transform: scale(1) translate(-50%, -50%);
            transform-origin: 0% 0%;
          }
          .bottom-left-rectangle {
            bottom: 0;
            left: 0;
            transform: scale(1) translate(-50%, 50%);
            transform-origin: 0% 100%;
          }
          .top-right-circle {
            top: 14%;
            right: 14%;
            transform: scale(1) translate(50%, -50%);
            transform-origin: 100% 0%;
          }
          .bottom-right-circle {
            bottom: 14%;
            right: 14%;
            transform: scale(1) translate(50%, 50%);
            transform-origin: 100% 100%;
          }
          .top-left-circle {
            top: 14%;
            left: 14%;
            transform: scale(1) translate(-50%, -50%);
            transform-origin: 0% 0%;
          }
          .bottom-left-circle {
            bottom: 14%;
            left: 14%;
            transform: scale(1) translate(-50%, 50%);
            transform-origin: 0% 100%;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Badge);
