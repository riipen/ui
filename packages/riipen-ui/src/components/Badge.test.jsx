import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Badge from "./Badge";

describe("<Badge>", () => {
  it("renders without errors", () => {
    let error;
    try {
      mount(<Badge />);
    } catch (theError) {
      error = theError;
    }
    expect(error).toEqual(undefined);
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Badge().type.defaultProps;

      const wrapper = mount(<Badge />);

      expect(wrapper.find("Badge").props().anchor.horizontal).toEqual(
        defaultProps.anchor.horizontal
      );
      expect(wrapper.find("Badge").props().anchor.vertical).toEqual(
        defaultProps.anchor.vertical
      );
      expect(wrapper.find("Badge").props().color).toEqual(defaultProps.color);
      expect(wrapper.find("Badge").props().component).toEqual(
        defaultProps.component
      );
      expect(wrapper.find("Badge").props().max).toEqual(defaultProps.max);
      expect(wrapper.find("Badge").props().overlap).toEqual(
        defaultProps.overlap
      );
      expect(wrapper.find("Badge").props().showZero).toEqual(
        defaultProps.showZero
      );
      expect(wrapper.find("Badge").props().size).toEqual(defaultProps.size);
      expect(wrapper.find("Badge").props().variant).toEqual(
        defaultProps.variant
      );

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe("anchor prop", () => {
    it("sets anchor class name with valid custom anchor", () => {
      const anchor = { horizontal: "left", vertical: "bottom" };

      const wrapper = mount(<Badge anchor={anchor} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass("bottom-left-rectangle")
      ).toEqual(true);
    });

    it("gives an error when given an invalid anchor", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const anchor = { horizontal: "side", vertical: "up" };

      mount(<Badge anchor={anchor} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>child</div>;

      const wrapper = mount(<Badge>{child}</Badge>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classVariant = ["newClass"];

      const wrapper = mount(<Badge classes={classVariant} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass(classVariant[0])
      ).toEqual(true);
    });
  });

  describe("color prop", () => {
    it("sets color class name with a valid color", () => {
      const color = "muted";

      const wrapper = mount(<Badge color={color} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass(color)
      ).toEqual(true);
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Badge color="blue" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("component prop", () => {
    it("sets component with valid custom component", () => {
      const component = "a";

      const wrapper = mount(<Badge component={component} />);

      expect(wrapper.find(component)).toHaveLength(1);
      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .name()
      ).toEqual(component);
    });

    it("gives an error when given an invalid component", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const component = "abcdefg";

      mount(<Badge component={component} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("content prop", () => {
    it("badge renders when content is 0 and showZero is false", () => {
      const content = 0;

      const wrapper = mount(<Badge content={content} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass("hidden")
      ).toEqual(true);
      expect(Number(wrapper.text())).toEqual(content);
    });

    it("badge renders when content is 0 and showZero is true", () => {
      const content = 0;
      const showZero = true;

      const wrapper = mount(<Badge content={content} showZero={showZero} />);

      expect(Number(wrapper.find("Badge").text())).toEqual(content);
    });

    it("badge renders when content is greater than 0 and showZero is true", () => {
      const content = 1;
      const showZero = true;

      const wrapper = mount(<Badge content={content} showZero={showZero} />);

      expect(Number(wrapper.find("Badge").text())).toEqual(content);
    });
  });

  describe("max prop", () => {
    it("displays max value given content greater than max value", () => {
      const max = 9;
      const content = 11;

      const wrapper = mount(<Badge max={max} content={content} />);

      expect(wrapper.text()).toEqual(`${max}+`);
    });

    it("displays content value given content less than max value", () => {
      const max = 9;
      const content = 8;

      const wrapper = mount(<Badge max={max} content={content} />);

      expect(Number(wrapper.text())).toEqual(content);
    });
  });

  describe("overlap prop", () => {
    it("sets overlap class name with valid custom overlap", () => {
      const anchor = { horizontal: "left", vertical: "top" };
      const overlap = "circle";

      const wrapper = mount(<Badge anchor={anchor} overlap={overlap} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass("top-left-circle")
      ).toEqual(true);
    });

    it("gives an error when given an invalid overlap", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Badge overlap="square" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("showZero prop", () => {
    it("badge renders when showZero prop value is true", () => {
      const showZero = true;

      const wrapper = mount(<Badge showZero={showZero} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass("hidden")
      ).toEqual(false);
    });
  });

  describe("size prop", () => {
    it("sets size class name with a valid size", () => {
      const size = "small";
      const variant = "standard";

      const wrapper = mount(<Badge size={size} variant={variant} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass(size)
      ).toEqual(true);
    });

    it("does not set size class name when non standard variant prop used", () => {
      const size = "small";
      const variant = "dot";

      const wrapper = mount(<Badge size={size} variant={variant} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass(size)
      ).toEqual(false);
    });

    it("gives an error when given an invalid size", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const size = "large";

      mount(<Badge size={size} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("variant prop", () => {
    it("sets variant class name with valid custom variant", () => {
      const variant = "dot";

      const wrapper = mount(<Badge variant={variant} />);

      expect(
        wrapper
          .find("Badge")
          .childAt(0)
          .childAt(0)
          .hasClass(variant)
      ).toEqual(true);
    });

    it("throws an error with with invalid variant", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Badge variant="circle" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });
});
