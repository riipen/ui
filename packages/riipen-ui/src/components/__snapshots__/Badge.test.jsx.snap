// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge> renders correct snapshot 1`] = `
<withClasses(Badge)>
  <Badge
    anchor={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    classes={
      Array [
        "riipen",
        "riipen-badge",
      ]
    }
    color="primary"
    component="span"
    max={99}
    overlap="rectangle"
    showZero={false}
    size="medium"
    variant="standard"
  >
    <span
      className="jsx-1015140608 root"
    >
      <span
        className="jsx-1015140608 badge hidden top-right-rectangle primary standard medium riipen riipen-badge"
      />
    </span>
    <JSXStyle
      dynamic={
        Array [
          10,
          "#373737",
          "Roboto, Helvetica, Arial, sans-serif",
          500,
          "10px",
          "14px",
          "12px",
          20,
          20,
          10,
          10,
          "#1f71ae",
          "#fff",
          "#f58632",
          "#fff",
          "#4cc88d",
          "#fff",
          "#e95353",
          "#fff",
          "#cccbcb",
          "#fff",
        ]
      }
      id="2138813230"
    />
  </Badge>
</withClasses(Badge)>
`;
